frappe/model/base_document.py

```diff
@@ -384,6 +384,10 @@ def get_valid_dict(
 			if ignore_nulls and not is_virtual_field and value is None:
 				continue
 
+			# If the docfield is not nullable - set a default non-null value
+			if value is None and getattr(df, "not_nullable", False):
+				value = ""
```

Comment:
This should be handled much earlier? Also defaults depend on fieldtype.

Reply:
A bit confused about this, initially I did set it up to get the default value dynamically - e76c806fed40bc59c288e0937f41be1c00c76cfa

But from what I can understand, its only the varchar types that require this, correct?